src/main/java/io/cos/cas/osf/web/flow/login/OsfPrincipalFromNonInteractiveCredentialsAction.java

```diff
@@ -627,35 +648,79 @@ private OsfApiInstitutionAuthenticationResult notifyOsfApiOfInstnAuthnSuccess(
-                        responseString
-                );
-                throw new InstitutionSsoFailedException("OSF API failed to process CAS request");
-            }
-            return new OsfApiInstitutionAuthenticationResult(username, institutionId);
         } catch (final IOException e) {
-            LOGGER.error(
-                    "[OSF API] Notify Remote Principal Authenticated Failed: Communication Error - {}",
-                    e.getMessage()
-            );
+            LOGGER.error("[OSF API] Notify Remote Principal Authenticated Failed: Communication Error - {}", e.getMessage());
             throw new InstitutionSsoFailedException("Communication Error between OSF CAS and OSF API");
         }
+        // CAS expects OSF API to return HTTP 204 OK with no content if authentication succeeds
+        if (statusCode == HttpStatus.SC_NO_CONTENT) {
+            LOGGER.info("[OSF API] Notify Remote Principal Authenticated Passed: institution={}, username={}", institutionId, username);
+            return new OsfApiInstitutionAuthenticationResult(username, institutionId);
+        }
+        // Handler unexpected exceptions (i.e. any status other than 403)
+        if (statusCode != HttpStatus.SC_FORBIDDEN) {
+            LOGGER.error("[OSF API] Notify Remote Principal Authenticated Failed: Unexpected Failure - statusCode={}", statusCode);
+            throw new InstitutionSsoFailedException("OSF API failed to process CAS request");
+        }
+        // CAS expects OSF API to return HTTP 403 FORBIDDEN with error details if authentication fails.
+        String responseRaw;
+        try {
+            HttpEntity entity = httpResponse.getEntity();
+            responseRaw = EntityUtils.toString(entity);
+        } catch (final IOException | ParseException e) {
+            LOGGER.error("[OSF API] Notify Remote Principal Authenticated Failed: Entity String Parse Error - {}", e.getMessage());
+            throw new InstitutionSsoFailedException("CAS fails to parse OSF API error response");
+        }
+        // Handle failures due to denied selective SSO
+        try {
+            final JsonObject responseJson = JsonParser.parseString(responseRaw).getAsJsonObject();
+            final JsonArray errorList = responseJson.getAsJsonArray("errors");
+            for (final JsonElement error : errorList) {
+                if (!error.isJsonObject()) {
+                    LOGGER.warn("[OSF API] Unexpected API Response Format: error is not a JSON object");
+                    continue;
+                }
+                if (!((JsonObject) error).has("detail")) {
+                    LOGGER.warn("[OSF API] Unexpected API Response Format: missing key \"detail\" in the error object");
+                    continue;
+                }
+                final String errorDetail = ((JsonObject) error).get("detail").getAsString();
+                if (OsfApiPermissionDenied.INSTITUTION_SELECTIVE_SSO_FAILURE.getId().equals(errorDetail)) {
```

Comment:
As mentioned, this new exception response handling supports other customized error messages which can be added as needed in the future.